utils/ort/src/main/kotlin/storage/S3FileStorage.kt

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2023 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.utils.ort.storage
+
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.io.InputStream
+import java.net.URI
+
+import org.apache.commons.compress.compressors.xz.XZCompressorInputStream
+import org.apache.commons.compress.compressors.xz.XZCompressorOutputStream
+import org.apache.logging.log4j.kotlin.logger
+
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider
+import software.amazon.awssdk.core.sync.RequestBody
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.GetObjectRequest
+import software.amazon.awssdk.services.s3.model.HeadObjectRequest
+import software.amazon.awssdk.services.s3.model.NoSuchKeyException
+import software.amazon.awssdk.services.s3.model.PutObjectRequest
+import software.amazon.awssdk.services.s3.model.S3Exception
+
+/**
+ * A [FileStorage] that stores files in an AWS S3 bucket [bucketName]. The [read] and [exists] operations are
+ * blocking, but the [write] operation is asynchronous unless a [customEndpoint] is provided. Contents are compressed
+ * before store if the [compression] flag is set to true.
+ */
+@Suppress("SwallowedException")
+class S3FileStorage(
+
+    /** The AWS access key */
+    private val accessKeyId: String? = null,
+
+    /** The AWS region to be used. */
+    private val awsRegion: String? = null,
+
+    /** The name of the S3 bucket used to store files in. */
+    private val bucketName: String,
+
+    /** Whether to use compression for storing files or not. Defaults to true. */
+    private val compression: Boolean = false,
+
+    /** Custom endpoint to perform AWS API Requests */
+    private val customEndpoint: String? = null,
+
+    /** The AWS secret for the access key. */
+    private val secretAccessKey: String? = null
+
+) : FileStorage {
+
+    val s3: S3Client? by lazy {
```

Comment:
Let's name the variable `s3Client` for clarity.

Reply:
Also, can this really ever be null? I acknowledge this comes from a Java platform call that doesn't support explicit non-nullability. But according to the docs the builder should never return `null`, or?